### ISSUE TITLE
Adds a transaction system

### ISSUE THREAD

**Issue by Tpt** [Tpt/oxigraph#19]
Oxigraph should expose a transaction system allowing to get ACID properties of operations consisting of multiples triples additions and/or removals. RocksDB supports transactions so it should not be too hard to implement.

**Comment by Tpt** [comment]
Done for the "write-only" part.

**Comment by ktk** [comment]
Related topics from SPARQL 1.2 CWG:
* [Multi-request transaction support in the SPARQL protocol](https://github.com/w3c/sparql-12/issues/83)
* [Relaxation of transaction requirements](https://github.com/w3c/sparql-12/issues/60)

**Comment by Tpt** [comment]
This is blocked by Sled for the read part: https://github.com/spacejam/sled/issues/1143

**Comment by Tpt** [comment]
Releated to #144. The storage system choice impacts a lot the availlability of this feature.

**Comment by Tpt** [comment]
I have just released [v0.3.0-beta.1](https://github.com/oxigraph/oxigraph/releases/tag/v0.3.0-beta.1) that provides a complete read/write transaction system. It enforces the "repeatable read" isolation level: the store only exposes changes that have been "committed" (i.e. no partial writes) and the exposed state does not change for the complete duration of a read operation (e.g. a SPARQL query) or a read/write operation (e.g. a SPARQL update).

**Comment by ktk** [comment]
Great, I suppose this also applies on Graph Store Protocol? As in as long as the write is not finished, it will neither replace the current data nor show up in a query?

**Comment by Tpt** [comment]
> Great, I suppose this also applies on Graph Store Protocol?
Yes, exactly. All Oxigraph operations are now transactional except bulk loading. This applies also to the SPARQL Protocol and Graph Store Protocol implementations. Bulk loading is not available from HTTP at the moment so all operations done using HTTP requests are benefiting the "repeatable read"  transactional guarantees.
